hsds/domain_crawl.py

```diff
@@ -0,0 +1,275 @@
+##############################################################################
+# Copyright by The HDF Group.                                                #
+# All rights reserved.                                                       #
+#                                                                            #
+# This file is part of HSDS (HDF5 Scalable Data Service), Libraries and      #
+# Utilities.  The full HSDS copyright notice, including                      #
+# terms governing use, modification, and redistribution, is contained in     #
+# the file COPYING, which can be found at the root of the source code        #
+# distribution tree.  If you do not have access to this file, you may        #
+# request a copy from help@hdfgroup.org.                                     #
+##############################################################################
+#
+# domain crawler
+#
+
+import asyncio
+
+from aiohttp.web_exceptions import HTTPServiceUnavailable, HTTPConflict, HTTPBadRequest
+from aiohttp.web_exceptions import HTTPInternalServerError, HTTPNotFound, HTTPGone
+
+
+from .util.idUtil import getCollectionForId, getDataNodeUrl
+from .util.httpUtil import http_post, http_put
+
+from .servicenode_lib import getObjectJson
+from . import hsds_logger as log
+
+
+class DomainCrawler:
+    def __init__(
+        self,
+        app,
+        objs,
+        action="get_obj",
+        params=None,  
+        max_tasks=40,
+        max_objects_limit=0,
+    ):
+        log.info(f"DomainCrawler.__init__  root_id: {len(objs)} objs")
+        self._app = app
+        self._action = action
+        self._max_objects_limit = max_objects_limit
+        self._params = params
+        self._max_tasks = max_tasks
+        self._q = asyncio.Queue()
+        self._obj_dict = {}
+        self.seen_ids = set()
+        if not objs:
+            log.error("no objs for crawler to crawl!")
+            raise ValueError()
+
+        for obj_id in objs:
+            log.debug(f"adding {obj_id} to the queue")
+            self._q.put_nowait(obj_id)
+        if isinstance(objs, dict):
+            self._objs = objs
+        else:
+            self._objs = None
+
+
+    async def get_attributes(self, obj_id, attr_names):
+        # get the given attributes for the obj_id
```

Comment:
`get the given attributes for the obj_id` -> `write the given attributes to the obj_id`